src/az_ulib_ustream_forward/az_ulib_ustream_forward.c

```diff
@@ -0,0 +1,161 @@
+      _Pragma("clang diagnostic ignored \"-Wincompatible-pointer-types-discards-qualifiers\"")
+#define IGNORE_MEMCPY_TO_NULL _Pragma("GCC diagnostic push")
+#define RESUME_WARNINGS _Pragma("clang diagnostic pop")
+#elif defined(__GNUC__)
+#define IGNORE_POINTER_TYPE_QUALIFICATION \
+  _Pragma("GCC diagnostic push") _Pragma("GCC diagnostic ignored \"-Wdiscarded-qualifiers\"")
+#define IGNORE_MEMCPY_TO_NULL _Pragma("GCC diagnostic push")
+#define RESUME_WARNINGS _Pragma("GCC diagnostic pop")
+#else
+#define IGNORE_POINTER_TYPE_QUALIFICATION __pragma(warning(push));
+#define IGNORE_MEMCPY_TO_NULL \
+  __pragma(warning(push));  \
+  __pragma(warning(suppress: 6387));
+#define RESUME_WARNINGS __pragma(warning(pop));
+#endif // __clang__
+
+static az_result concrete_flush(
+    az_ulib_ustream_forward* ustream_forward,
+    az_ulib_flush_callback flush_callback, 
+    az_ulib_callback_context flush_callback_context);
+static az_result concrete_read(
+    az_ulib_ustream_forward* ustream_forward,
+    uint8_t* const buffer,
+    size_t buffer_length,
+    size_t* const size);
+static size_t concrete_get_size(
+    az_ulib_ustream_forward* ustream_forward);
+static az_result concrete_dispose(
+    az_ulib_ustream_forward* ustream_forward);
+static const az_ulib_ustream_forward_interface api
+    = { concrete_flush, concrete_read,  concrete_get_size, concrete_dispose };
+
+static az_result concrete_flush(
+    az_ulib_ustream_forward* ustream_forward,
+    az_ulib_flush_callback flush_callback, 
+    az_ulib_callback_context flush_callback_context)
+{
+  // precondition checks
+  _az_PRECONDITION_NOT_NULL(ustream_forward);
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+  _az_PRECONDITION_NOT_NULL(flush_callback);
+
+  // get size of data
+  size_t buffer_size = concrete_get_size(ustream_forward);
+
+  // point to data
+  const uint8_t* buffer = (const uint8_t*)ustream_forward->_internal.ptr;
+
+  // invoke callback
+  (*flush_callback)(buffer, buffer_size, flush_callback_context);
+
+  return AZ_OK;
+}
+
+static az_result concrete_read(
+    az_ulib_ustream_forward* ustream_forward,
+    uint8_t* const buffer,
+    size_t buffer_length,
+    size_t* const size)
+{
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+  _az_PRECONDITION_NOT_NULL(buffer);
+  _az_PRECONDITION(buffer_length > 0);
+  _az_PRECONDITION_NOT_NULL(size);
+
+  az_result result;
+
+  if (ustream_forward->_internal.inner_current_position >= ustream_forward->_internal.length)
+  {
+    *size = 0;
+    result = AZ_ULIB_EOF;
+  }
+  else
+  {
+    size_t remain_size
+        = ustream_forward->_internal.length - (size_t)ustream_forward->_internal.inner_current_position;
+    *size = (buffer_length < remain_size) ? buffer_length : remain_size;
+
+    /**
+     * MSVC thinks `buffer` could be '0'. Because we make sure `buffer` is not `NULL` with a 
+     * precondition, this is not a concern. We thus ignore this compiler error.
```

Comment:
This is not the real reason. In debug mode, the precondition macros check if the buffer is `NULL`, so the compiler will not throw any warning for that. However, in release mode, the test in the precondition macro is removed, so the compiler will complain about this function using a pointer without test it, and the compiler is right. 

We made the decision to do not test it in release because it is an developer error, and return an error will not help in this situation, so the best thing to do here is let the `memcpy` crash with this `NULL` pointer.

Reply:
I'm confused. does this mean I should disable it or toggle it with `AZ_NO_PRECONDITION_CHECKING`? Or does it mean that the reason is that the only way this can fail is in release mode and at that point returning an error will not help?

Reply:
"With precondition off (release mode), compiler will complain that destination pointer could be `NULL`. However, this is expected behavior if the user has followed the SDK guidelines [URL]."

Reply:
https://azure.github.io/azure-sdk/clang_design.html#:~:text=Supported%20platforms%20%20%20%20Operating%20System%20,%20%20gcc-7.x%20%205%20more%20rows%20

Reply:
https://azure.github.io/azure-sdk/clang_design.html#pre-conditions